### DISCUSSION TITLE
Paths referenced by plugins.luasnip.from{Lua,Snipmate,Vscode}.*.paths are removed by `nix store gc`

### DISCUSSION THREAD

**Discussion by cezdro** [nix-community/nixvim#2162] [Q&A]
I configured the snippet path using the relative path like in [the example](https://nix-community.github.io/nixvim/plugins/luasnip/fromLua.html#pluginsluasnipfromlua). However, at one point the snippets stopped working and I figured that it happened because Nix thought the directory was only needed at a build time and could now be GC'd (I have `nix.gc.automatic` enabled). Here is the snippet from the generated `init.lua`:

```lua
require("luasnip.loaders.from_lua").lazy_load({
    ["paths"] = "/nix/store/1wm0qsxvri2pdxd67sxa569inz4pl385-f5jcwka5bpj542475ib1kgfplgnb9x8n-source/modules/home/nvim/plugins/luasnip",
})

require("luasnip.loaders.from_snipmate").lazy_load({
    ["paths"] = "/nix/store/1wm0qsxvri2pdxd67sxa569inz4pl385-f5jcwka5bpj542475ib1kgfplgnb9x8n-source/modules/home/nvim/plugins/luasnip",
})

require("luasnip.loaders.from_vscode").lazy_load({
    ["paths"] = "/nix/store/1wm0qsxvri2pdxd67sxa569inz4pl385-f5jcwka5bpj542475ib1kgfplgnb9x8n-source/modules/home/nvim/plugins/luasnip",
})
```

I don't know how Nix decides whether the referenced path should be kept or not. I *think* that if I used `mkDerivation` the directory wouldn't be treated as a garbage, but I don't think that such thing should be necessary in Nixvim config.

I wrote this in Discussions, because I don't know if this is my fault and if it can be improved in Nixvim.

**Comment by MattSturgeon** [comment]
I don't have a full grasp on this, but it's definitely unfortunate if paths referenced in a nixvim config aren't treated as persistent runtime dependencies...

Reading [this nix pill](https://nixos.org/guides/nix-pills/09-automatic-runtime-dependencies), it seems nix usually figures it out automagically. However I wonder if, bucause the path is to a `-source` type derivation, perhaps things are treated differently? 

In a nixpkgs style derivation all runtime/installed files will have been copied to the builder's `$out` path.

Looking at how nixos handles paths in (for example) its `etc` module, it seems they ensure all referenced paths are _copied_ (not symlinked) into the final derivation. We do the same thing in our `extraFiles` module: https://github.com/nix-community/nixvim/blob/9c11b54065a554d9976089cfc8a08dee35cabcaa/modules/top-level/files/default.nix#L70-L92

My first thought is that _maybe_ any module allowing path type options should be coercing them into a derivation owned by the module implementing the option, but I feel like we'd need to discuss the issue with some experts to get a better understanding of the situation.

As an initial workaround, does doing something like `builtins.toFile "filename" (builtins.readFile ./srcFile)` help at all?

**Comment by cezdro** [comment]
I ended up doing this:

```nix
{ pkgs, ... }: let
  snippets = pkgs.stdenv.mkDerivation {
    name = "snippets";

    src = ./.;

    buildCommand = ''
      mkdir $out
      cp -r $src/*.{json,lua,snippets} $out
    '';
  };
in {
  programs.nixvim = {
    plugins.luasnip = {
      enable = true;

      fromLua = [{paths = "${snippets}";}];
      fromSnipmate = [{paths = "${snippets}";}];
      fromVscode = [{paths = "${snippets}";}];
    };
  };
}
```

The "snippets" derivation isn't removed on `nix store gc`. Anyway, I still think this is something that Nixvim could do automatically.

**Comment by cezdro** [comment]
I created an issue from this discussion: #2179
